src/Orchestration/Adapter/DockerAPI.php

```diff
@@ -0,0 +1,380 @@
+            switch ($stream) { // only 1 or 2, as set while creating exec 
+                case 1:
+                    $packed = pack('C*', ...\array_slice($rawStream, 8));
+                    $stdout .= $packed;
+                    break;
+                case 2:
+                    $packed = pack('C*', ...\array_slice($rawStream, 8));
+                    $stderr .= $packed;
+                    break;
+            }
+            return strlen($str); // must return full frame from callback
+        };
+        \curl_setopt($ch, CURLOPT_WRITEFUNCTION, $callback);
+
+
+        \curl_setopt($ch, CURLOPT_TIMEOUT, $timeout);
+        $result = \curl_exec($ch);
+        $responseCode = \curl_getinfo($ch, CURLINFO_RESPONSE_CODE);
+
+        if(curl_errno($ch))
+        {
+            if (\curl_errno($ch) == CURLE_OPERATION_TIMEOUTED) {
+                throw new TimeoutException('Curl Error: ' . curl_error($ch));
+            } else {
+                throw new DockerAPIException('Curl Error: ' . curl_error($ch));
+            }
+        }
+
+        \curl_close($ch);
+
+        return [
+            'response' => $result,
+            'code' => $responseCode,
+            'stdout' => $stdout,
+            'stderr' => $stderr
+        ];
+    }
+
+    /**
+     * Pull Image
+     * 
+     * @param string $image
+     * 
+     * @return bool
+     */
+    public function pull(string $image): bool
+    {
+        $result = $this->call('http://localhost/images/create', 'POST', \http_build_query([
+            'fromImage' => $image
+        ]), [
+            'X-Registry-Auth' => $this->registryAuth
+        ]);
+
+        if ($result['code'] !== 200 && $result['code'] !== 204) {
+            $data = json_decode($result['response'], true);
+            if (isset($data['message'])) {
+                throw new DockerAPIException('Failed to pull container: ' . $data['message']);
+            } else {
+                throw new DockerAPIException('Failed to pull container: Internal Docker Error');
+            }
+            return false;
+        } else {
+            return true;
+        }
+    }
+
+    /**
+     * List Containers
+     * @param array<string, string> $filters
+     *
+     * @return Container[]
+     */
+    public function list(array $filters = []): array
+    {
+        $filtersSorted = [];
+        array_walk($filters,
+            function(string $value, string $key) use (&$filtersSorted){
+                $filtersSorted[$key] = [$value];
+            });
+
+        $body = [
+            'all' => true,
+            'filters' => empty($filtersSorted) ? new stdClass() : json_encode($filtersSorted)
+        ];
+
+        $result = $this->call('http://localhost/containers/json'.'?'.\http_build_query($body), 'GET');
+
+        $list = [];
+
+        if ($result['code'] !== 200) {
+            throw new DockerAPIException($result['response']);
+        }
+
+        \array_map(function(array $value) use (&$list) {
```

Comment:
Here we are using `array_map` without returning any value in the callback - so not really necessary.